src/Storage/Hydration/EventHydrator.php

```diff
@@ -0,0 +1,167 @@
+            );
+
+        }
+
+        foreach ($properties as $property) {
+            $this->setEventProperty($event, $property, $data);
+        }
+
+        return $event;
+    }
+
+
+
+    /**
+     * @return mixed[]
+     */
+    public function dehydrate(Event $event) : array
+    {
+        $data = [];
+
+        $reflection = new ReflectionClass($event);
+        $properties = $this->getProperties($reflection);
+
+        foreach ($properties as $property) {
+            $data[$property->getName()] = $this->getEventPropertyValue($event, $property);
+        }
+
+        $data['eventType'] = $event::eventType();
+
+        return $data;
+    }
+
+
+    /**
+     * @return mixed
+     */
+    private function getEventPropertyValue(Event $event, ReflectionProperty $property)
+    {
+        // First check to see if we have a getter method. If we do, use that to get the value instead.
+        $method = $this->findGetterForProperty($property);
+
+        if ($method) {
+            return $method->invoke($event);
+        }
+
+        $property->setAccessible(true);
+
+        return $property->getValue($event);
+    }
+
+
+    /**
+     * @param mixed[] $data
+     */
+    private function setEventProperty(Event $event, ReflectionProperty $property, array $data) : void
+    {
+        if (!isset($data[$property->getName()])) {
+            return;
+        }
+
+        $property->setAccessible(true);
+        $property->setValue($event, $data[$property->getName()]);
+    }
+
+
+    private function findGetterForProperty(ReflectionProperty $property) : ?ReflectionMethod
+    {
+        $reflection = $property->getDeclaringClass();
+        $methodName = sprintf('get%s', ucfirst($property->getName()));
+        $method     = $this->getMethodOrNull($reflection, $methodName);
+
+        // If we can't find a method explicitly prefixed with 'get' then try finding a property matching the parameter
+        // name exactly.
+        if (!$method) {
+            $method = $this->getMethodOrNull($reflection, $property->getName());
+        }
+
+        // Only use the method if it looks like a getter.
+        if ($method && $method->isPublic() && $method->hasReturnType() && $method->getNumberOfParameters() === 0) {
+            return $method;
+        }
+
+        return null;
+    }
+
+
+    private function getMethodOrNull(ReflectionClass $reflection, string $name) : ?ReflectionMethod
+    {
+        try {
+            return $reflection->getMethod($name);
+        } catch (ReflectionException $exception) {
+            return null;
+        }
+    }
+
+
+    /**
+     * @return ReflectionProperty[]
+     */
+    private function getProperties(ReflectionClass $reflection) : array
+    {
+        $properties = $reflection->getProperties();
+        $parent     = $reflection->getParentClass();
+
+        while ($parent) {
+            $properties = array_merge($properties, $parent->getProperties(ReflectionProperty::IS_PRIVATE));
```

Comment:
Should this throw the case where there is >1 property with the same name? In that case we can't reliably know which one to set.

Reply:
Yep that would be clearer for consumers. We could also try handling it in a more advanced way by prefixing these properties with something that indicates what level in the class hierarchy they came from.

Reply:
I think for more complex cases, something like https://github.com/schmittjoh/metadata would be a good option.